snuba/datasets/errors_replacer.py

```diff
@@ -38,6 +43,72 @@ class ReplacerState(Enum):
     ERRORS = "errors"
 
 
+@dataclass(frozen=True)
+class NeedsFinal:
+    pass
+
+
+@dataclass(frozen=True)
+class ExcludeGroups:
+    group_ids: Sequence[int]
+
+
+QueryTimeFlags = Union[NeedsFinal, ExcludeGroups]
+
+
+class Replacement(ReplacementBase):
+    @abstractmethod
+    def get_query_time_flags(self) -> Optional[QueryTimeFlags]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_project_id(self) -> int:
+        raise NotImplementedError()
+
+
+EXCLUDE_GROUPS = object()
+NEEDS_FINAL = object()
+LegacyQueryTimeFlags = Union[Tuple[object, int], Tuple[object, int, Any]]
+
+
+@dataclass(frozen=True)
+class LegacyReplacement(Replacement):
+    # XXX: For the group_exclude message we need to be able to run a
+    # replacement without running any query.
+    count_query_template: Optional[str]
+    insert_query_template: Optional[str]
+    query_args: Mapping[str, Any]
+    query_time_flags: LegacyQueryTimeFlags
+
+    def get_project_id(self) -> int:
+        return self.query_time_flags[1]
+
+    def get_query_time_flags(self) -> Optional[QueryTimeFlags]:
+        if self.query_time_flags[0] == NEEDS_FINAL:
+            return NeedsFinal()
+
+        if self.query_time_flags[0] == EXCLUDE_GROUPS:
+            return ExcludeGroups(group_ids=self.query_time_flags[2])  # type: ignore
```

Comment:
This should go away when we stop using legacy query time flags, right ?

Reply:
yes the entire class should go away